base/linalg/lapack.jl

```diff
@@ -1780,6 +1780,74 @@ for (orglq, orgqr, ormlq, ormqr, gemqrt, elty) in
                 A
             end
         end
+        # SUBROUTINE DORGQL( M, N, K, A, LDA, TAU, WORK, LWORK, INFO )
+        # *     .. Scalar Arguments ..
+        #       INTEGER            INFO, K, LDA, LWORK, M, N
+        # *     .. Array Arguments ..
+        #       DOUBLE PRECISION   A( LDA, * ), TAU( * ), WORK( * )
+        function orgql!(A::StridedMatrix{$elty}, tau::Vector{$elty}, k::Integer = length(tau))
+            chkstride1(A)
+            m = size(A, 1)
+            n = min(m, size(A, 2))
+            if k > n
+                throw(DimensionMismatch("Invalid number of reflectors: k = $k should be <= n = $n"))
+            end
+            work  = Array($elty, 1)
+            lwork = BlasInt(-1)
+            info  = Array(BlasInt, 1)
+            for i in 1:2
+                ccall(($(blasfunc(orgql)), liblapack), Void,
+                      (Ptr{BlasInt}, Ptr{BlasInt}, Ptr{BlasInt}, Ptr{$elty},
+                       Ptr{BlasInt}, Ptr{$elty}, Ptr{$elty}, Ptr{BlasInt}, Ptr{BlasInt}),
+                      &m, &n, &k, A,
+                      &max(1,stride(A,2)), tau, work, &lwork,
+                      info)
+                @lapackerror
+                if lwork < 0
+                    lwork = BlasInt(real(work[1]))
+                    work = Array($elty, lwork)
+                end
+            end
+            if n<size(A,2)
+                A[:,1:n]
+            else
+                A
+            end
+        end
+        # SUBROUTINE DORGRQ( M, N, K, A, LDA, TAU, WORK, LWORK, INFO )
+        # *     .. Scalar Arguments ..
+        #       INTEGER            INFO, K, LDA, LWORK, M, N
+        # *     .. Array Arguments ..
+        #       DOUBLE PRECISION   A( LDA, * ), TAU( * ), WORK( * )
+        function orgrq!(A::StridedMatrix{$elty}, tau::Vector{$elty}, k::Integer = length(tau))
+            chkstride1(A)
+            m = size(A, 1)
+            n = min(m, size(A, 2))
+            if k > n
+                throw(DimensionMismatch("Invalid number of reflectors: k = $k should be <= n = $n"))
+            end
+            work  = Array($elty, 1)
+            lwork = BlasInt(-1)
+            info  = Array(BlasInt, 1)
+            for i in 1:2
+                ccall(($(blasfunc(orgrq)), liblapack), Void,
+                      (Ptr{BlasInt}, Ptr{BlasInt}, Ptr{BlasInt}, Ptr{$elty},
+                       Ptr{BlasInt}, Ptr{$elty}, Ptr{$elty}, Ptr{BlasInt}, Ptr{BlasInt}),
+                      &m, &n, &k, A,
+                      &max(1,stride(A,2)), tau, work, &lwork,
+                      info)
+                @lapackerror
+                if lwork < 0
+                    lwork = BlasInt(real(work[1]))
+                    work = Array($elty, lwork)
+                end
+            end
+            if n<size(A,2)
+                A[:,1:n]
```

Comment:
I think type stability is more important than avoiding the copy. Another option is to always return `A` even though the last columns are garbage. That behavior should be clear from the LAPACK documentation.